sdks/python/apache_beam/transforms/ptransform.py

```diff
@@ -97,29 +100,112 @@ def visit(self, node, replacements):
       return self.visit_nested(node, replacements)
 
 
+# Caches to allow for materialization of values when executing a pipeline
+# in-process, in eager mode.  This cache allows the same _MaterializedResult
+# object to be accessed and used despite Runner API round-trip serialization.
+_pipeline_materialization_cache = {}
+_pipeline_materialization_lock = threading.Lock()
+
+
+def _allocate_materialized_pipeline(pipeline):
+  with _pipeline_materialization_lock:
+    pipeline_id = id(pipeline)
+    _pipeline_materialization_cache[pipeline_id] = {}
+
```

Comment:
<!--thread_id:cc_164549409_t; commit:f44a011a9c23103ec0bf0c047e4b3415452efb2c; resolved:1-->
<!--section:context-quote-->
> **robertwb** wrote:
> Thanks, though I think we still need to handle the case of fork() is being used to spawn workers (maybe check PID as well as presence in dict?). 
> 
> Is this global rather than, say, methods on pipeline itself, just to support passing through pickling?

<!--section:body-->
I've added `pid` to the cache key.

Yes, this is global just to support passing through pickling.